File: source/extensions/isaacsim.robot_motion.experimental.motion_generation/python/impl/_logging.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Library-owned logging configuration for motion generation."""

from __future__ import annotations

import threading
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from isaacsim.common.logging import Logger

_DEFAULT_LOGGING_CHANNEL = "isaacsim.robot_motion.motion_gen"


class _LoggingState:
    """Manage the logging channel and lazily create its logger."""

    def __init__(self) -> None:
        self._channel = _DEFAULT_LOGGING_CHANNEL
        self._configured = False
        self._logger: Logger | None = None
        self._lock = threading.Lock()

    def configure_channel(self, channel: str) -> None:
        """Configure the channel before the logger is first used."""
        if not channel:
            raise ValueError("The logging channel must not be empty")

        with self._lock:
            if self._logger is not None and self._logger.channel != channel:
                raise RuntimeError("The logging channel cannot be changed after the logger has been created")
            if self._configured and self._channel != channel:
                raise RuntimeError("The logging channel has already been configured")
            self._channel = channel
            self._configured = True

    def get_logger(self) -> Logger:
        """Return the library logger, creating it on first use."""
        with self._lock:
            if self._logger is None:
                from isaacsim.common.logging import Logger

                self._logger = Logger(self._channel)
            return self._logger


class _LazyLogger:
    """Forward attribute access to the library logger without eagerly creating it."""

    def __getattr__(self, name: str) -> object:
        return getattr(_LOGGING_STATE.get_logger(), name)


_LOGGING_STATE = _LoggingState()

if TYPE_CHECKING:
    log: Logger
else:
    log = _LazyLogger()


def configure_logging_channel(channel: str) -> None:
    """Configure the motion generation logging channel before its first use."""
    _LOGGING_STATE.configure_channel(channel)


def get_logger() -> Logger:
    """Return the lazily initialized motion generation logger."""
    return _LOGGING_STATE.get_logger()

File: source/libraries/isaacsim/common/logging/tests/c/Logging.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "isaacsim/common/logging/Logging.h"

#include "isaacsim/common/logging/LoggingHost.h"

#include <doctest/doctest.h>

#include <cstddef>
#include <cstdint>

TEST_SUITE("C API")
{
    TEST_CASE("Global configuration accepts the legacy structure size")
    {
        IsaacSimCommonLoggingGlobalConfig config = ISAACSIM_COMMON_LOGGING_GLOBAL_CONFIG_INIT;
        config.fields = ISAACSIM_COMMON_LOGGING_CONFIG_ELAPSED_TIME;
        config.elapsedTime = ISAACSIM_COMMON_LOGGING_ELAPSED_TIME_MILLISECONDS;

        CHECK(isaacsimCommonLoggingConfigureGlobal(&config) == ISAACSIM_COMMON_LOGGING_CONFIGURE_SUCCESS);

        config.structSize = offsetof(IsaacSimCommonLoggingGlobalConfig, channelConfigs);
        CHECK(isaacsimCommonLoggingConfigureGlobal(&config) == ISAACSIM_COMMON_LOGGING_CONFIGURE_SUCCESS);
    }

    TEST_CASE("Channel enablement can be overridden and restored")
    {
        constexpr const char* kChannel = "isaacsim.test.c";
        IsaacSimCommonLoggingChannelConfig channel = ISAACSIM_COMMON_LOGGING_CHANNEL_CONFIG_INIT;
        channel.channel = kChannel;
        channel.enabledBehavior = ISAACSIM_COMMON_LOGGING_CHANNEL_SETTING_OVERRIDE;
        channel.enabled = 0;

        IsaacSimCommonLoggingGlobalConfig config = ISAACSIM_COMMON_LOGGING_GLOBAL_CONFIG_INIT;
        config.fields = ISAACSIM_COMMON_LOGGING_CONFIG_CHANNELS;
        config.channelConfigs = &channel;
        config.channelConfigCount = 1;

        REQUIRE(isaacsimCommonLoggingConfigureGlobal(&config) == ISAACSIM_COMMON_LOGGING_CONFIGURE_SUCCESS);
        ISAACSIM_COMMON_LOGGING_ERROR(kChannel, "disabled channel record");

        channel.enabledBehavior = ISAACSIM_COMMON_LOGGING_CHANNEL_SETTING_INHERIT;
        CHECK(isaacsimCommonLoggingConfigureGlobal(&config) == ISAACSIM_COMMON_LOGGING_CONFIGURE_SUCCESS);
    }

    TEST_CASE("Logging macros submit C API records")
    {
        constexpr const char* kChannel = "isaacsim.test.c";
        ISAACSIM_COMMON_LOGGING_VERBOSE(kChannel, "verbose record");
        ISAACSIM_COMMON_LOGGING_INFO(kChannel, "info record");
        ISAACSIM_COMMON_LOGGING_WARNING(kChannel, "isaacsim-c-api-backend-probe");
        ISAACSIM_COMMON_LOGGING_ERROR(kChannel, "error record");
    }

    TEST_CASE("Report writes to standard output and submits to Carbonite")
    {
        ISAACSIM_COMMON_LOGGING_REPORT("isaacsim.test.c", "isaacsim-c-report-probe");
    }

    TEST_CASE("Invalid supplemental logging input is ignored")
    {
        constexpr const char* kChannel = "isaacsim.test.c";
        isaacsimCommonLoggingWrite(ISAACSIM_COMMON_LOGGING_LOG_INFO, nullptr, "message", nullptr, nullptr, 0);
        isaacsimCommonLoggingWrite(ISAACSIM_COMMON_LOGGING_LOG_INFO, kChannel, nullptr, nullptr, nullptr, 0);
        isaacsimCommonLoggingWrite(
            static_cast<IsaacSimCommonLoggingLogLevel>(99), kChannel, "message", nullptr, nullptr, 0);
        isaacsimCommonLoggingReport(nullptr, "message", nullptr, nullptr, 0);
        isaacsimCommonLoggingReport(kChannel, nullptr, nullptr, nullptr, 0);
    }

    TEST_CASE("Invalid global configuration is rejected")
    {
        CHECK(isaacsimCommonLoggingConfigureGlobal(nullptr) == ISAACSIM_COMMON_LOGGING_CONFIGURE_INVALID_ARGUMENT);

        IsaacSimCommonLoggingChannelConfig channel = ISAACSIM_COMMON_LOGGING_CHANNEL_CONFIG_INIT;
        channel.channel = "isaacsim.test.c";
        IsaacSimCommonLoggingGlobalConfig config = ISAACSIM_COMMON_LOGGING_GLOBAL_CONFIG_INIT;

        config.fields = UINT64_C(1) << 63;
        CHECK(isaacsimCommonLoggingConfigureGlobal(&config) == ISAACSIM_COMMON_LOGGING_CONFIGURE_INVALID_ARGUMENT);

        config.fields = ISAACSIM_COMMON_LOGGING_CONFIG_CHANNELS;
        config.channelConfigs = &channel;
        config.channelConfigCount = 1;
        channel.enabledBehavior = 99;
        CHECK(isaacsimCommonLoggingConfigureGlobal(&config) == ISAACSIM_COMMON_LOGGING_CONFIGURE_INVALID_ARGUMENT);

        channel.enabledBehavior = ISAACSIM_COMMON_LOGGING_CHANNEL_SETTING_INHERIT;
        config.channelConfigs = nullptr;
        CHECK(isaacsimCommonLoggingConfigureGlobal(&config) == ISAACSIM_COMMON_LOGGING_CONFIGURE_INVALID_ARGUMENT);

        config.channelConfigs = &channel;
        channel.structSize = 0;
        CHECK(isaacsimCommonLoggingConfigureGlobal(&config) == ISAACSIM_COMMON_LOGGING_CONFIGURE_INVALID_ARGUMENT);
    }

    TEST_CASE("Flush drains the process backend")
    {
        isaacsimCommonLoggingFlush();
    }

    TEST_CASE("Host integration rejects invalid ownership input")
    {
        IsaacSimCommonLoggingHostToken token = 0;
        CHECK(isaacsimCommonLoggingAttachCarboniteHost(nullptr, &token) == ISAACSIM_COMMON_LOGGING_HOST_INVALID_ARGUMENT);
        CHECK(isaacsimCommonLoggingAttachCarboniteHost(reinterpret_cast<void*>(1), nullptr) ==
              ISAACSIM_COMMON_LOGGING_HOST_INVALID_ARGUMENT);
        CHECK(isaacsimCommonLoggingDetachCarboniteHost(0) == ISAACSIM_COMMON_LOGGING_HOST_INVALID_ARGUMENT);
    }
}
